agent/submatview/store.go

```diff
@@ -0,0 +1,220 @@
+	// last request for that entry has been terminated. It is a field on the struct
+	// so that it can be patched in tests without need a lock.
+	idleTTL time.Duration
+}
+
+type entry struct {
+	materializer *Materializer
+	expiry       *ttlcache.Entry
+	stop         func()
+	// requests is the count of active requests using this entry. This entry will
+	// remain in the store as long as this count remains > 0.
+	requests int
+}
+
+func NewStore(logger hclog.Logger) *Store {
+	return &Store{
+		logger:     logger,
+		byKey:      make(map[string]entry),
+		expiryHeap: ttlcache.NewExpiryHeap(),
+		idleTTL:    20 * time.Minute,
+	}
+}
+
+// Run the expiration loop until the context is cancelled.
+func (s *Store) Run(ctx context.Context) {
+	for {
+		s.lock.RLock()
+		timer := s.expiryHeap.Next()
+		s.lock.RUnlock()
+
+		select {
+		case <-ctx.Done():
+			timer.Stop()
+			return
+		case <-s.expiryHeap.NotifyCh:
+			timer.Stop()
+			continue
+
+		case <-timer.Wait():
+			s.lock.Lock()
+
+			he := timer.Entry
+			s.expiryHeap.Remove(he.Index())
+
+			e := s.byKey[he.Key()]
+
+			// Only stop the materializer if there are no active requests.
+			if e.requests == 0 {
+				e.stop()
+				delete(s.byKey, he.Key())
+			}
+
+			s.lock.Unlock()
+		}
+	}
+}
+
+// TODO: godoc
+type Request interface {
+	cache.Request
+	NewMaterializer() (*Materializer, error)
+	Type() string
+}
+
+// Get a value from the store, blocking if the store has not yet seen the
+// req.Index value.
+// See agent/cache.Cache.Get for complete documentation.
+func (s *Store) Get(ctx context.Context, req Request) (Result, error) {
+	info := req.CacheInfo()
+	key, e, err := s.getEntry(req)
+	if err != nil {
+		return Result{}, err
+	}
+	defer s.releaseEntry(key)
+
+	ctx, cancel := context.WithTimeout(ctx, info.Timeout)
+	defer cancel()
+
+	result, err := e.materializer.getFromView(ctx, info.MinIndex)
+
+	// TODO: does context.DeadlineExceeded need to be translated into a nil error
+	// to match the old interface?
+
+	return result, err
+}
+
+// Notify the updateCh when there are updates to the entry identified by req.
+// See agent/cache.Cache.Notify for complete documentation.
+//
+// Request.CacheInfo().Timeout is ignored because it is not really relevant in
+// this case. Instead set a deadline on the context.
+func (s *Store) Notify(
+	ctx context.Context,
+	req Request,
+	correlationID string,
+	updateCh chan<- cache.UpdateEvent,
+) error {
+	info := req.CacheInfo()
+	key, e, err := s.getEntry(req)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		defer s.releaseEntry(key)
+
+		index := info.MinIndex
+		for {
+			result, err := e.materializer.getFromView(ctx, index)
+			switch {
+			case ctx.Err() != nil:
+				return
+			case err != nil:
+				// TODO: cache.Notify sends errors on updateCh, should this do the same?
+				// It seems like only fetch errors would ever get sent along and eventually
+				// logged, so sending may not provide any benefit here.
```

Comment:
Making a note of this TODO. This is a small change in behaviour from `Cache.Notify` which might be significant.